src/Database/PostgreSQL/LibPQ.hsc

```diff
@@ -463,10 +469,24 @@ pollHelper poller connection =
 --
 -- Note that the 'Connection' must not be used again after 'finish'
 -- has been called.
+--
+-- When this function returns, is is guaranteed that @PQfinish()@
+-- was called and the connection is closed.
 finish :: Connection
        -> IO ()
-finish (Conn fp _) =
+finish (Conn fp _ finishedVar) =
     do finalizeForeignPtr fp
+       -- This MVar is necessary because we unfortunately use `newForeignPtr`
```

Comment:
I'm confused. The [`finalizeForeignPtr`](https://hackage.haskell.org/package/base-4.16.1.0/docs/GHC-ForeignPtr.html#v:finalizeForeignPtr) docs say

> Causes the finalizers associated with a foreign pointer to be run immediately. The foreign pointer must not be used again after this function is called. If the foreign pointer does not support finalizers, this is a no-op.

Is that wrong, and it's actually just *immediately scheduled to be run*?

Reply:
The way I understand it so far is as follows:

* "Normal" `ForeignPtr` finalizers are function pointers to C functions. This allows some extra guarantees, most importantly that they can be run _during_ a GC (which is impossible for a finalizer function writtten in Haskell), and independent of any Haskell thread scheduling (thus truly "run immediately").
* The Haskell-`IO` based finalizer are a special case, for which the "run immediately` of `finalizeForeignPtr` doesn't apply. I take that from
  > [`addForeignPtrConcFinalizer`](https://hackage.haskell.org/package/base-4.16.1.0/docs/GHC-ForeignPtr.html#v:addForeignPtrConcFinalizer) - This is a variant of `addForeignPtrFinalizer,` where the finalizer is an arbitrary `IO` action. When it [?] is invoked, the finalizer will run in a new thread.

  I understand the `it` in this sentence as "the finalization of this ForeignPtr", including as triggered by `finalizeForeignPtr`.

It is possible that my understanding is wrong, that an explicit call to `finalizeForeignPtr` will override the `When it is invoked, the finalizer will run in a new thread` logic, instead of running it on the current thread. But the docs don't really say that.

I think it would be very useful for somebody that understands the ForeignPtr implementation very well (maybe @ezyang from [this post](http://blog.ezyang.com/2010/07/managing-foreign-pointers-effectively/)?) to chime in and explain which doc's rule takes precedence when `finalizeForeignPtr` is run on a `addForeignPtrConcFinalizer`. If the promptness guarantee of `finalizeForeignPtr` wins, that'd be great and make `Foreign.Concurrent.newForeignPtr` less dangerous, and we could update the docs accordingly.

Reply:
I think @bgamari is another person who may know. Until there's input from GHC developers I'd rather not merge this. Either way, GHC docs should be clarified. I don't want to guess (i.e. I want to do this right).

Reply:
@phadej Yes, good approach.

Linking your upstream GHC question: https://gitlab.haskell.org/ghc/ghc/-/issues/21420

Reply:
`finalizeForeignPtr` does run finalizers immediately. Specifically, looking at `GHC.ForeignPtr.foreignPtrFinalizer` you will see that there are two varieties of `ForeignPtr` finalizers:

* `HaskellFinalizers`, which is just a list of `IO ()` actions which are immediately run
* `CFinalizers`, which is a `Weak#` with an attached C finalization function. `Weak#`s can have two types of finalizers: zero or more C finalizers and a single Haskell finalizer. To run these  `foreignPtrFinalizer` uses the `finalizeWeak#` primop, which runs the C finalizers and returns Haskell finalizer. `foreignPtrFinalize` then runs the returned Haskell finalizer if it is valid.

So, `finalizeForeignPtr` should finalize immediately in all cases. However, @nh2 is correct in the case of finalizers run by the garbage collector: finalizers are run concurrently with the mutator after a major GC has determined that a `Weak#` can be finalized. This is handled by `Weak.c:scheduleFinalizers`.

Aside: The fact that both `ForeignPtr`s and `Weak#` both support both C and Haskell finalizers is oddly redundant; I suspect that the  reasons for this redundancy are strictly historical accident. It might be worth cleaning this up.

Reply:
So this PR is unnecessary?

Reply:
I think so, closing. Thanks all!